runtime/codegen/graph.go

```diff
@@ -0,0 +1,70 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package codegen
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"regexp"
+	"sort"
+)
+
+// Component graph edges are embedded in the generated binary as
+// specially formatted strings. These strings can be extracted from
+// the binary to get the communication graph without having to execute
+// the binary.
+//
+// Each edge is represented by a string fragment that looks like:
+// ⟦checksum:wEaVeReDgE:src→dst⟧
+//
+// checksum is the first 8 bytes of the hex encoding of the SHA-256 of
+// the string "src→dst" and src and dst are the fully qualified
+// component type names.
+
+// MakeEdgeString returns a string that should be emitted into generated
+// code to represent an edge from src to dst.
+func MakeEdgeString(src, dst string) string {
+	return fmt.Sprintf("⟦%s:wEaVeReDgE:%s→%s⟧\n", checksum(src, dst), src, dst)
```

Comment:
I thought this would be hacky, but the addition of `wEaVeReDgE` makes it totally legit! :rofl: